File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.cc
#include <iostream>
#include <stack>
#include <algorithm>

#define FastIO ios::sync_with_stdio(false), cin.tie(nullptr), cout.tie(nullptr);

using namespace std;

int N;
int arr[100000];

int main() {
    FastIO;

    while (true) {
        cin >> N;
        if (N == 0) break;

        for (int i = 0; i < N; i++) {
            cin >> arr[i];
        }

        stack<int> s;
        int idx = 0;

        long long res = 0;

        while (idx < N) {
            if (s.empty() || arr[s.top()] <= arr[idx]) {
                s.push(idx++);
            }
            else {
                int top = s.top();
                s.pop();

                long long area = (long long)arr[top] * (s.empty() ? idx : idx - s.top() - 1);
                res = max(res, area);
            }
        }

        while (!s.empty()) {
            int top = s.top();
            s.pop();

            long long area = (long long)arr[top] * (s.empty() ? idx : idx - s.top() - 1);
            res = max(res, area);
        }

        cout << res << "\n";
    }

    return 0;
}

File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.py
def largest_rectangle(histogram):
    stack = []
    max_area = 0
    index = 0

    while index < len(histogram):
        if not stack or histogram[stack[-1]] <= histogram[index]:
            stack.append(index)
            index += 1
        else:
            top_of_stack = stack.pop()
            area = (histogram[top_of_stack] *
                    ((index - stack[-1] - 1) if stack else index))
            max_area = max(max_area, area)

    while stack:
        top_of_stack = stack.pop()
        area = (histogram[top_of_stack] *
                ((index - stack[-1] - 1) if stack else index))
        max_area = max(max_area, area)

    return max_area

while True:
    inputs = list(map(int, input().split()))
    n = inputs[0]
    if n == 0:
        break
    histogram = inputs[1:]
    print(largest_rectangle(histogram))
